### PR TITLE
fix bug  when page and popup are closed

### PR DESCRIPTION
**Pull request by jefleponot**
When you click on a link that close the current page => CasperJS crashes (casperJS can't work without a main page)
When you are on a popup and you click to close the page => you have to return to the main page else it occurs troubles.

### CODE REVIEW

**Comment by jefleponot** [comment]
@mickaelandrieu ,@bguerin, @mlalpho 

To conclude with popup bugs, here is a last PR that gives an optimize and readable code.

Thank for review

File: modules/casper.js [hunk 1]
```diff
@@ -404,11 +404,6 @@ Casper.prototype.captureSelector = function captureSelector(targetFile, selector
 Casper.prototype.checkStep = function checkStep(self, onComplete) {
     "use strict";
 
-    try{
-        var test = self.page.loadInProgress || self.navigationRequested || self.page.browserInitializing ; 
-    } catch(e){
-        self.page = self.mainPage;
-    }
```
Comment (jefleponot) [inline modules/casper.js]: I remove this code because try and catch options slow down main loop

File: modules/casper.js [hunk 2]
```diff
@@ -720,10 +715,7 @@ Casper.prototype.evaluate = function evaluate(fn, context) {
     if (!utils.isFunction(fn) && !utils.isString(fn)) { // phantomjs allows functions defs as string
         throw new CasperError("evaluate() only accepts functions or strings");
     }
-    // closed page check
-    if (this.page.closed) {
-        return null;
-    }
```
Comment (jefleponot) [inline modules/casper.js]: this code was never used : I had to remove it

File: modules/casper.js [hunk 3]
```diff
@@ -2585,7 +2577,9 @@ Casper.prototype.withPopup = function withPopup(popupInfo, then) {
 Casper.prototype.newPage = function newPage() {
     "use strict";
     this.checkStarted();
-    this.page.close();
+    if (this.page !== null) {
+        this.page.close();
+    }
```
Comment (jefleponot) [inline modules/casper.js]: add a test when page had been yet closed

File: modules/casper.js [hunk 4]
```diff
@@ -2644,10 +2638,15 @@ function createPage(casper) {
     mainPage.isPopup = false;
 
     var onClosing = function onClosing(closedPopup) {
-        if (this.isPopup) {
+        if (closedPopup.isPopup) {
+            if (casper.page.id === closedPopup.id) {
+                casper.page = casper.mainPage;
+            }
```
Comment (jefleponot) [inline modules/casper.js]: checks if popup is being used when closing

File: modules/casper.js [hunk 5]
```diff
@@ -2644,10 +2638,15 @@ function createPage(casper) {
...
             casper.popups.clean();
-            closedPopup.closed = true;
             casper.emit('popup.closed', closedPopup);
+        } else {
+            casper.page = null;
+            casper.newPage();
```
Comment (jefleponot) [inline modules/casper.js]: casperJS can't run without any page

File: modules/casper.js [hunk 6]
```diff
@@ -2729,6 +2728,7 @@ function createPage(casper) {
     
     var onPageCreated = function onPageCreated(page) {
         page.isPopup = (typeof page.isPopup === "undefined") ? true : false;
+        page.id = new Date().getTime();
```
Comment (jefleponot) [inline modules/casper.js]: adds an uniq id for popups

**Comment by mickaelandrieu** [comment]
thanks!

**Pull request merged**